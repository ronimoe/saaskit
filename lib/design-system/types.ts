// Design System Type Definitions
// Centralized type definitions for all design system components and utilities

// === COMPONENT VARIANTS ===

export type GlassVariant = 'primary' | 'secondary' | 'accent' | 'floating' | 'subtle'
export type GlassDepth = 'subtle' | 'medium' | 'deep' | 'floating'
export type GlassGlow = 'none' | 'subtle' | 'medium' | 'strong'
export type GlassSize = 'sm' | 'md' | 'lg' | 'xl'

export type MagneticStrength = 'subtle' | 'light' | 'medium' | 'strong'
export type MagneticDistance = 'close' | 'medium' | 'far' | 'extreme'

export type AnimationDurationType = 'fast' | 'normal' | 'slow' | 'slower'
export type AnimationEasingType = 'linear' | 'ease' | 'ease-in' | 'ease-out' | 'ease-in-out' | 'smooth' | 'spring' | 'bounce'

export type ParticleCount = 'minimal' | 'light' | 'medium' | 'heavy' | 'extreme'
export type ParticleSize = 'tiny' | 'small' | 'medium' | 'large' | 'huge'
export type ParticleSpeed = 'slow' | 'medium' | 'fast' | 'rapid'

export type GridCellType = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
export type GridGap = 'tight' | 'normal' | 'relaxed' | 'loose'

// === THEME TYPES ===

export interface ThemeColors {
  primary: string
  secondary: string
  accent: string
  success: string
  warning: string
  error: string
  info: string
}

export interface BrandConfiguration {
  primary: string
  secondary: string
  accent: string
  logo?: string
  name?: string
}

export interface PerformanceToggles {
  glassmorphism: boolean
  animations: boolean
  particleEffects: boolean
  magneticEffects: boolean
  highContrast: boolean
}

export interface AccessibilityOptions {
  reducedMotion: boolean
  highContrast: boolean
  screenReaderOptimized: boolean
  keyboardNavigation: boolean
}

// === COMPONENT PROPS ===

export interface BaseComponentProps {
  className?: string
  children?: React.ReactNode
  'data-testid'?: string
}

export interface GlassComponentProps extends BaseComponentProps {
  variant?: GlassVariant
  size?: GlassSize
  depth?: GlassDepth
  glow?: GlassGlow
  interactive?: boolean
}

export interface MagneticComponentProps extends BaseComponentProps {
  magnetic?: boolean
  magneticStrength?: number
  magneticDistance?: number
  magneticGlow?: boolean
}

export interface AnimatedComponentProps extends BaseComponentProps {
  animate?: boolean
  animationDuration?: AnimationDurationType
  animationEasing?: AnimationEasingType
  animationDelay?: number
  stagger?: boolean
  staggerDelay?: number
}

export interface ParticleComponentProps extends BaseComponentProps {
  particleCount?: number
  particleColor?: string
  particleSize?: number
  particleSpeed?: number
  connectionDistance?: number
  mouseInteraction?: boolean
}

export interface LayoutComponentProps extends BaseComponentProps {
  asymmetric?: boolean
  cellType?: GridCellType
  diagonal?: boolean
  floating?: boolean
  perspective?: boolean
}

// === INTERACTION TYPES ===

export interface InteractionState {
  isHover: boolean
  isFocus: boolean
  isActive: boolean
  isPressed: boolean
  isDisabled: boolean
}

export interface MotionConfig {
  respectReducedMotion: boolean
  enableHardwareAcceleration: boolean
  enableGPUAcceleration: boolean
  enableTransforms: boolean
}

export interface GestureConfig {
  enableMagnetic: boolean
  enableParallax: boolean
  enableTilt: boolean
  enableHover: boolean
  enableTouch: boolean
}

// === LAYOUT TYPES ===

export interface BreakpointConfig {
  xs: number
  sm: number
  md: number
  lg: number
  xl: number
  '2xl': number
}

export interface SpacingScale {
  xs: string
  sm: string
  md: string
  lg: string
  xl: string
  '2xl': string
  '3xl': string
  '4xl': string
  '5xl': string
}

export interface TypographyScale {
  xs: string
  sm: string
  base: string
  lg: string
  xl: string
  '2xl': string
  '3xl': string
  '4xl': string
  '5xl': string
  '6xl': string
  '7xl': string
  '8xl': string
  '9xl': string
}

// === ANIMATION TYPES ===

export interface AnimationKeyframes {
  from: Record<string, string | number>
  to: Record<string, string | number>
  [key: string]: Record<string, string | number>
}

export interface TransitionConfig {
  property: string | string[]
  duration: string | number
  timingFunction: string
  delay?: string | number
}

export interface KeyframeAnimation {
  name: string
  keyframes: AnimationKeyframes
  duration: string | number
  timingFunction?: string
  delay?: string | number
  iterationCount?: number | 'infinite'
  direction?: 'normal' | 'reverse' | 'alternate' | 'alternate-reverse'
  fillMode?: 'none' | 'forwards' | 'backwards' | 'both'
}

// === COMPOSITION TYPES ===

export interface CompositionLayer {
  type: 'glass' | 'magnetic' | 'particle' | 'layout' | 'theme'
  props: Record<string, unknown>
  priority: number
  conditional?: boolean
}

export interface CompositionStack {
  layers: CompositionLayer[]
  baseClasses: string
  responsive?: boolean
  accessible?: boolean
  performanceOptimized?: boolean
}

// === UTILITY TYPES ===

export type ResponsiveValue<T> = T | {
  base?: T
  sm?: T
  md?: T
  lg?: T
  xl?: T
  '2xl'?: T
}

export type ConditionalValue<T> = T | {
  default: T
  hover?: T
  focus?: T
  active?: T
  disabled?: T
}

export type VariantValue<T> = T | {
  [variant: string]: T
}

// === CONFIGURATION TYPES ===

export interface DesignSystemConfig {
  theme: {
    colors: ThemeColors
    brand?: BrandConfiguration
    breakpoints: BreakpointConfig
    spacing: SpacingScale
    typography: TypographyScale
  }
  features: {
    glassmorphism: boolean
    magneticEffects: boolean
    particleAnimations: boolean
    nonTraditionalLayouts: boolean
    dynamicTheming: boolean
    gamification: boolean
  }
  performance: PerformanceToggles
  accessibility: AccessibilityOptions
  motion: MotionConfig
  gestures: GestureConfig
}

export interface ComponentLibrary {
  glass: {
    GlassCard: React.ComponentType<GlassComponentProps>
    MagneticGlassCard: React.ComponentType<GlassComponentProps & MagneticComponentProps>
  }
  interactive: {
    UnfoldableFeatureCard: React.ComponentType<BaseComponentProps>
    FeatureConnections: React.ComponentType<BaseComponentProps>
    AnimatedProductMockup: React.ComponentType<AnimatedComponentProps>
  }
  animation: {
    ParticleBackground: React.ComponentType<ParticleComponentProps>
  }
  layout: {
    AsymmetricGrid: React.ComponentType<LayoutComponentProps>
    DiagonalSection: React.ComponentType<LayoutComponentProps>
    FloatingElement: React.ComponentType<LayoutComponentProps>
  }
}

// === HOOK TYPES ===

export interface MagneticHookOptions {
  strength?: number
  smoothing?: number
  maxDistance?: number
  resetDuration?: number
  enabled?: boolean
}

export interface ThemeHookReturn {
  theme: string
  setTheme: (theme: string) => void
  config: DesignSystemConfig
  updateConfig: (config: Partial<DesignSystemConfig>) => void
  colors: ThemeColors
  brand?: BrandConfiguration
}

export interface AnimationHookOptions {
  duration?: AnimationDurationType
  easing?: AnimationEasingType
  delay?: number
  autoPlay?: boolean
  loop?: boolean
  direction?: 'normal' | 'reverse' | 'alternate'
}

// === EVENT TYPES ===

export interface DesignSystemEvent {
  type: 'theme-change' | 'performance-toggle' | 'accessibility-change' | 'animation-complete'
  payload: Record<string, unknown>
  timestamp: number
}

export interface InteractionEvent {
  type: 'hover' | 'focus' | 'click' | 'touch' | 'magnetic-activate' | 'particle-interact'
  target: HTMLElement
  coordinates?: { x: number; y: number }
  force?: number
  timestamp: number
}

// === VALIDATION TYPES ===

export interface ValidationRule<T> {
  required?: boolean
  type?: 'string' | 'number' | 'boolean' | 'object' | 'array'
  validator?: (value: T) => boolean
  message?: string
}

export type ValidationSchema<T> = {
  [K in keyof T]?: ValidationRule<T[K]>
}

export interface ValidationResult {
  isValid: boolean
  errors: Array<{
    field: string
    message: string
    value: unknown
  }>
}

// === BROWSER SUPPORT TYPES ===

export interface BrowserSupport {
  backdropFilter: string[]
  cssGrid: string[]
  customProperties: string[]
  intersectionObserver: string[]
  resizeObserver: string[]
}

export interface FeatureDetection {
  backdropFilter: boolean
  cssGrid: boolean
  customProperties: boolean
  intersectionObserver: boolean
  resizeObserver: boolean
  touchEvents: boolean
  pointerEvents: boolean
  webGL: boolean
}

// === EXPORT COLLECTIONS ===

export type DesignSystemComponent = keyof ComponentLibrary[keyof ComponentLibrary]

export type AllVariants = GlassVariant | MagneticStrength | AnimationEasing | ParticleCount | GridCellType

export type AllSizes = GlassSize | ParticleSize

export type AllColors = keyof ThemeColors

export type AllSpacing = keyof SpacingScale

export type AllTypography = keyof TypographyScale 