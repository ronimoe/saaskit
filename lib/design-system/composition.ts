// Component Composition Utilities
// Helper functions for combining design system components and effects

import { type ClassValue, clsx } from 'clsx'
import { twMerge } from 'tailwind-merge'

// === UTILITY FUNCTIONS ===

/**
 * Utility function to combine class names with tailwind-merge
 */
export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

/**
 * Generate responsive classes based on breakpoints
 */
export function responsive(classes: Record<string, string>) {
  return Object.entries(classes)
    .map(([breakpoint, className]) => {
      if (breakpoint === 'base') return className
      return `${breakpoint}:${className}`
    })
    .join(' ')
}

// === GLASS COMPOSITION ===

export interface GlassCompositionProps {
  variant?: 'primary' | 'secondary' | 'accent' | 'floating' | 'subtle'
  depth?: 'subtle' | 'medium' | 'deep' | 'floating'
  glow?: 'none' | 'subtle' | 'medium' | 'strong'
  magnetic?: boolean
  magneticStrength?: number
  className?: string
}

/**
 * Compose glass effect classes with magnetic effects
 */
export function composeGlassEffects({
  variant = 'primary',
  depth = 'medium',
  glow = 'none',
  magnetic = false,
  className = ''
}: GlassCompositionProps) {
  return cn(
    // Base glass styles
    'backdrop-blur-sm bg-background/80 border border-border/50',
    
    // Variant styles
    {
      'bg-white/10 dark:bg-black/10 border-white/20 dark:border-white/10': variant === 'primary',
      'bg-secondary/10 border-secondary/20': variant === 'secondary',
      'bg-accent/10 border-accent/20': variant === 'accent',
      'bg-white/5 dark:bg-black/5 border-white/10 dark:border-white/5': variant === 'subtle',
      'bg-white/15 dark:bg-black/15 border-white/30 dark:border-white/20 shadow-2xl': variant === 'floating',
    },
    
    // Depth styles
    {
      'shadow-md': depth === 'subtle',
      'shadow-lg': depth === 'medium', 
      'shadow-xl': depth === 'deep',
      'shadow-2xl drop-shadow-xl': depth === 'floating',
    },
    
    // Glow effects
    {
      'before:absolute before:inset-0 before:rounded-[inherit] before:border before:border-white/20': glow === 'subtle',
      'before:absolute before:inset-0 before:rounded-[inherit] before:border before:border-white/30 before:shadow-[inset_0_1px_0_0_rgba(255,255,255,0.1)]': glow === 'medium',
      'before:absolute before:inset-0 before:rounded-[inherit] before:border before:border-white/40 before:shadow-[inset_0_1px_0_0_rgba(255,255,255,0.2)]': glow === 'strong',
    },
    
    // Magnetic effects
    {
      'magnetic-element cursor-pointer transition-transform duration-200': magnetic,
    },
    
    className
  )
}

// === ANIMATION COMPOSITION ===

export interface AnimationCompositionProps {
  magnetic?: boolean
  particles?: boolean
  stagger?: boolean
  staggerDelay?: number
  hover?: boolean
  className?: string
}

/**
 * Compose animation classes for interactive elements
 */
export function composeAnimations({
  magnetic = false,
  particles = false,
  stagger = false,
  hover = true,
  className = ''
}: AnimationCompositionProps) {
  return cn(
    // Base animation styles
    'transition-all duration-300 ease-smooth',
    
    // Magnetic effects
    {
      'magnetic transform-gpu will-change-transform': magnetic,
    },
    
    // Stagger animations
    {
      'stagger-animation': stagger,
    },
    
    // Hover animations
    {
      'hover:scale-105 hover:shadow-lg hover:-translate-y-1': hover,
    },
    
    // Particle background container
    {
      'relative overflow-hidden': particles,
    },
    
    className
  )
}

// === LAYOUT COMPOSITION ===

export interface LayoutCompositionProps {
  asymmetric?: boolean
  cellType?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
  diagonal?: boolean
  floating?: boolean
  className?: string
}

/**
 * Compose layout classes for non-traditional layouts
 */
export function composeLayout({
  asymmetric = false,
  cellType,
  diagonal = false,
  floating = false,
  className = ''
}: LayoutCompositionProps) {
  return cn(
    // Base layout
    'relative',
    
    // Asymmetric grid
    {
      'asymmetric-grid': asymmetric,
    },
    
    // Grid cell types
    cellType && `grid-cell-${cellType}`,
    
    // Diagonal sections
    {
      'diagonal-section': diagonal,
    },
    
    // Floating elements
    {
      'floating-element absolute animate-float': floating,
    },
    
    className
  )
}

// === THEME COMPOSITION ===

export interface ThemeCompositionProps {
  brand?: boolean
  interactive?: boolean
  conditional?: boolean
  highContrast?: boolean
  className?: string
}

/**
 * Compose theme-aware classes
 */
export function composeTheme({
  brand = false,
  interactive = false,
  conditional = false,
  highContrast = false,
  className = ''
}: ThemeCompositionProps) {
  return cn(
    // Brand integration
    {
      'text-brand-primary bg-brand-primary border-brand-primary': brand,
    },
    
    // Interactive colors
    {
      'interactive-colors': interactive,
    },
    
    // Conditional features
    {
      'conditional-glassmorphism conditional-animation': conditional,
    },
    
    // High contrast mode
    {
      'high-contrast': highContrast,
    },
    
    className
  )
}

// === COMPLETE COMPOSITION ===

export interface CompleteCompositionProps {
  base?: string
  // Glass props
  variant?: 'primary' | 'secondary' | 'accent' | 'floating' | 'subtle'
  depth?: 'subtle' | 'medium' | 'deep' | 'floating'
  glow?: 'none' | 'subtle' | 'medium' | 'strong'
  magnetic?: boolean
  magneticStrength?: number
  // Animation props
  particles?: boolean
  stagger?: boolean
  staggerDelay?: number
  hover?: boolean
  // Layout props
  asymmetric?: boolean
  cellType?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
  diagonal?: boolean
  floating?: boolean
  // Theme props
  brand?: boolean
  interactive?: boolean
  conditional?: boolean
  highContrast?: boolean
  // Common
  className?: string
}

/**
 * Compose all design system effects into a single className
 */
export function composeComplete(props: CompleteCompositionProps) {
  const {
    base = '',
    className = '',
    variant, depth, glow, magnetic, magneticStrength,
    particles, stagger, staggerDelay, hover,
    asymmetric, cellType, diagonal, floating,
    brand, interactive, conditional, highContrast
  } = props
  
  return cn(
    base,
    composeGlassEffects({ variant, depth, glow, magnetic, magneticStrength }),
    composeAnimations({ magnetic, particles, stagger, staggerDelay, hover }),
    composeLayout({ asymmetric, cellType, diagonal, floating }),
    composeTheme({ brand, interactive, conditional, highContrast }),
    className
  )
}

// === PERFORMANCE UTILITIES ===

/**
 * Conditionally apply composition based on performance settings
 */
export function performanceAwareComposition(
  composition: string,
  conditions: {
    respectReducedMotion?: boolean
    respectGlassmorphismToggle?: boolean
    respectAnimationToggle?: boolean
  } = {}
) {
  let classes = composition
  
  // Add conditional classes based on performance toggles
  if (conditions.respectReducedMotion) {
    classes = `${classes} motion-safe:transition-all motion-reduce:transition-none`
  }
  
  if (conditions.respectGlassmorphismToggle) {
    classes = `${classes} conditional-glassmorphism`
  }
  
  if (conditions.respectAnimationToggle) {
    classes = `${classes} conditional-animation`
  }
  
  return classes
}

// === ACCESSIBILITY COMPOSITION ===

/**
 * Add accessibility-aware classes to compositions
 */
export function accessibleComposition(
  composition: string,
  options: {
    focusVisible?: boolean
    highContrast?: boolean
    screenReader?: boolean
  } = {}
) {
  return cn(
    composition,
    {
      'focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2': options.focusVisible,
      'contrast-more:border-contrast-more:bg-contrast-more:text-contrast': options.highContrast,
      'sr-only': options.screenReader,
    }
  )
} 