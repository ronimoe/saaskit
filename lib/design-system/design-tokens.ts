// Design Tokens
// Centralized design values for consistent styling across the system

// === SPACING TOKENS ===
export const SPACING = {
  // Base spacing scale (rem values)
  xs: '0.25rem',    // 4px
  sm: '0.5rem',     // 8px
  md: '0.75rem',    // 12px
  lg: '1rem',       // 16px
  xl: '1.5rem',     // 24px
  '2xl': '2rem',    // 32px
  '3xl': '3rem',    // 48px
  '4xl': '4rem',    // 64px
  '5xl': '6rem',    // 96px
  
  // SaaS Layout specific
  sidebar: '16rem',
  header: '3.5rem',
  content: '1.5rem',
} as const

// === TYPOGRAPHY TOKENS ===
export const TYPOGRAPHY = {
  fontFamilies: {
    sans: 'var(--font-geist-sans)',
    mono: 'var(--font-geist-mono)',
  },
  
  // Type scale
  sizes: {
    xs: '0.75rem',      // 12px
    sm: '0.875rem',     // 14px
    base: '1rem',       // 16px
    lg: '1.125rem',     // 18px
    xl: '1.25rem',      // 20px
    '2xl': '1.5rem',    // 24px
    '3xl': '1.875rem',  // 30px
    '4xl': '2.25rem',   // 36px
    '5xl': '3rem',      // 48px
    '6xl': '3.75rem',   // 60px
    '7xl': '4.5rem',    // 72px
    '8xl': '6rem',      // 96px
    '9xl': '8rem',      // 128px
  },
  
  lineHeights: {
    tight: '1.25',
    snug: '1.375',
    normal: '1.5',
    relaxed: '1.625',
    loose: '2',
  },
  
  fontWeights: {
    light: '300',
    normal: '400',
    medium: '500',
    semibold: '600',
    bold: '700',
    extrabold: '800',
  },
} as const

// === COLOR TOKENS ===
export const COLORS = {
  // Semantic colors
  semantic: {
    primary: 'var(--primary)',
    secondary: 'var(--secondary)',
    accent: 'var(--accent)',
    success: 'var(--success)',
    warning: 'var(--warning)',
    error: 'var(--destructive)',
    info: 'var(--info)',
  },
  
  // Brand colors (dynamic)
  brand: {
    primary: 'var(--brand-primary)',
    secondary: 'var(--brand-secondary)',
    accent: 'var(--brand-accent)',
    success: 'var(--brand-success)',
    warning: 'var(--brand-warning)',
    error: 'var(--brand-error)',
    info: 'var(--brand-info)',
  },
  
  // Surface colors
  surfaces: {
    background: 'var(--background)',
    foreground: 'var(--foreground)',
    card: 'var(--card)',
    popover: 'var(--popover)',
    muted: 'var(--muted)',
    border: 'var(--border)',
  },
  
  // Glass effect colors
  glass: {
    subtle: 'rgba(255, 255, 255, 0.03)',
    light: 'rgba(255, 255, 255, 0.08)',
    medium: 'rgba(255, 255, 255, 0.15)',
    strong: 'rgba(255, 255, 255, 0.25)',
  },
} as const

// === BORDER RADIUS TOKENS ===
export const RADIUS = {
  none: '0',
  sm: 'calc(var(--radius) - 4px)',
  md: 'calc(var(--radius) - 2px)',
  lg: 'var(--radius)',
  xl: 'calc(var(--radius) + 4px)',
  '2xl': 'calc(var(--radius) + 8px)',
  '3xl': 'calc(var(--radius) + 16px)',
  full: '9999px',
} as const

// === SHADOW TOKENS ===
export const SHADOWS = {
  // Standard shadows
  sm: '0 1px 2px 0 rgb(0 0 0 / 0.05)',
  md: '0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)',
  lg: '0 10px 15px -3px rgb(0 0 0 / 0.1), 0 4px 6px -4px rgb(0 0 0 / 0.1)',
  xl: '0 20px 25px -5px rgb(0 0 0 / 0.1), 0 8px 10px -6px rgb(0 0 0 / 0.1)',
  '2xl': '0 25px 50px -12px rgb(0 0 0 / 0.25)',
  
  // Glass shadows
  glass: {
    light: '0 4px 16px rgba(0, 0, 0, 0.05)',
    medium: '0 8px 32px rgba(0, 0, 0, 0.1)',
    strong: '0 12px 40px rgba(0, 0, 0, 0.15)',
    floating: '0 25px 50px rgba(0, 0, 0, 0.25)',
  },
  
  // Glow effects
  glow: {
    subtle: '0 0 20px var(--primary)',
    medium: '0 0 40px var(--primary)',
    strong: '0 0 60px var(--primary)',
  },
} as const

// === ANIMATION TOKENS ===
export const ANIMATIONS = {
  durations: {
    fast: '150ms',
    normal: '250ms',
    slow: '350ms',
    slower: '500ms',
  },
  
  easings: {
    linear: 'linear',
    ease: 'ease',
    'ease-in': 'ease-in',
    'ease-out': 'ease-out',
    'ease-in-out': 'ease-in-out',
    smooth: 'cubic-bezier(0.4, 0, 0.2, 1)',
    spring: 'cubic-bezier(0.175, 0.885, 0.32, 1.275)',
    bounce: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
  },
  
  delays: {
    none: '0ms',
    short: '100ms',
    medium: '200ms',
    long: '300ms',
    longer: '500ms',
  },
} as const

// === BREAKPOINTS ===
export const BREAKPOINTS = {
  xs: '480px',
  sm: '640px',
  md: '768px',
  lg: '1024px',
  xl: '1280px',
  '2xl': '1536px',
} as const

// === Z-INDEX SCALE ===
export const Z_INDEX = {
  dropdown: 1000,
  sticky: 1020,
  fixed: 1030,
  'modal-backdrop': 1040,
  modal: 1050,
  popover: 1060,
  tooltip: 1070,
  toast: 1080,
} as const

// === GLASS EFFECT TOKENS ===
export const GLASS_EFFECTS = {
  opacity: {
    subtle: 0.03,
    light: 0.08,
    medium: 0.15,
    strong: 0.25,
  },
  
  blur: {
    subtle: '8px',
    light: '10px',
    medium: '12px',
    deep: '16px',
    floating: '20px',
  },
  
  border: {
    subtle: 0.1,
    light: 0.2,
    medium: 0.3,
    strong: 0.4,
  },
} as const

// === MAGNETIC EFFECT TOKENS ===
export const MAGNETIC_EFFECTS = {
  strength: {
    subtle: 0.1,
    light: 0.2,
    medium: 0.3,
    strong: 0.5,
  },
  
  maxDistance: {
    close: 50,
    medium: 100,
    far: 150,
    extreme: 200,
  },
  
  smoothing: {
    instant: 0,
    fast: 0.1,
    medium: 0.2,
    smooth: 0.3,
  },
} as const

// === PARTICLE EFFECT TOKENS ===
export const PARTICLE_EFFECTS = {
  count: {
    minimal: 20,
    light: 40,
    medium: 60,
    heavy: 80,
    extreme: 100,
  },
  
  size: {
    tiny: 1,
    small: 2,
    medium: 3,
    large: 4,
    huge: 6,
  },
  
  speed: {
    slow: 0.1,
    medium: 0.3,
    fast: 0.5,
    rapid: 0.8,
  },
  
  connectionDistance: {
    close: 80,
    medium: 120,
    far: 160,
    extreme: 200,
  },
} as const

// === LAYOUT GRID TOKENS ===
export const GRID = {
  columns: {
    full: 12,
    tablet: 8,
    mobile: 1,
  },
  
  gaps: {
    tight: '0.5rem',
    normal: '1rem',
    relaxed: '1.5rem',
    loose: '2rem',
  },
  
  cellTypes: {
    1: { span: 4, rows: 2 },
    2: { span: 2, rows: 3 },
    3: { span: 6, rows: 1 },
    4: { span: 3, rows: 2 },
    5: { span: 5, rows: 3 },
    6: { span: 2, rows: 1 },
    7: { span: 4, rows: 1 },
    8: { span: 3, rows: 1 },
    9: { span: 3, rows: 2 },
  },
} as const

// === INTERACTIVE TOKENS ===
export const INTERACTIVE = {
  hover: {
    intensity: 1.1,
    duration: '200ms',
    easing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  },
  
  active: {
    intensity: 0.9,
    duration: '100ms',
    easing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  },
  
  focus: {
    intensity: 1.05,
    duration: '150ms',
    easing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  },
} as const

// === TOKEN COLLECTIONS ===
export const DESIGN_TOKENS = {
  spacing: SPACING,
  typography: TYPOGRAPHY,
  colors: COLORS,
  radius: RADIUS,
  shadows: SHADOWS,
  animations: ANIMATIONS,
  breakpoints: BREAKPOINTS,
  zIndex: Z_INDEX,
  glassEffects: GLASS_EFFECTS,
  magneticEffects: MAGNETIC_EFFECTS,
  particleEffects: PARTICLE_EFFECTS,
  grid: GRID,
  interactive: INTERACTIVE,
} as const

// === TYPE EXPORTS ===
export type SpacingToken = keyof typeof SPACING
export type ColorToken = keyof typeof COLORS.semantic
export type RadiusToken = keyof typeof RADIUS
export type ShadowToken = keyof typeof SHADOWS
export type AnimationDuration = keyof typeof ANIMATIONS.durations
export type AnimationEasing = keyof typeof ANIMATIONS.easings
export type Breakpoint = keyof typeof BREAKPOINTS 