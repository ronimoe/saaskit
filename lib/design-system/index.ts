// Design System Entry Point
// Centralized exports for all innovative UI components and utilities

// === COMPONENTS ===
// Glass Components (14.1)
export { GlassCard } from '@/components/ui/glass-card'
export { MagneticGlassCard } from '@/components/ui/magnetic-glass-card'

// Interactive Components (14.6)  
export { UnfoldableFeatureCard } from '@/components/ui/unfoldable-feature-card'
export { FeatureConnections } from '@/components/ui/feature-connections'
export { AnimatedProductMockup } from '@/components/ui/animated-product-mockup'

// Animation Components (14.3)
export { ParticleBackground } from '@/components/ui/particle-background'

// Layout Components (14.4)
// Note: Layout components are primarily CSS-based with utility classes

// === HOOKS ===
// Magnetic Effects Hook (14.3)
export { useMagneticEffect } from '@/lib/hooks/use-magnetic-effect'

// Theme Configuration Hook (14.2)
export { useThemeConfig } from '@/lib/hooks/use-theme-config'

// === UTILITIES ===
// Theme Configuration and Types (14.2)
export { 
  type ThemeConfig,
  type BrandColors,
  predefinedThemes,
  generateColorShades,
  adjustOklchLightness,
  createBrandPalette,
  validateThemeConfig,
  getThemeById,
  getThemesByMode
} from '@/lib/theme-config'

// Gamification Utilities (14.5)
export {
  type Achievement,
  type UserProgress,
  type UserLevel,
  calculateUserLevel,
  calculateUserXP,
  calculateProfileCompletion,
  getAvailableAchievements,
  checkAchievements,
  getAchievementRarityStyles,
  calculateUserProgress
} from '@/lib/gamification-utils'

// === DESIGN TOKENS ===
export * from '@/lib/design-system/design-tokens'

// === COMPOSITION UTILITIES ===
export * from '@/lib/design-system/composition'

// === TYPES ===
export * from '@/lib/design-system/types'

// === VERSION ===
export const DESIGN_SYSTEM_VERSION = '1.0.0'

// === FEATURE FLAGS ===
export const FEATURES = {
  GLASSMORPHISM: true,
  MAGNETIC_EFFECTS: true, 
  PARTICLE_ANIMATIONS: true,
  INTERACTIVE_CONNECTIONS: true,
  GAMIFICATION: true,
  NON_TRADITIONAL_LAYOUTS: true,
  DYNAMIC_THEMING: true,
  ACCESSIBILITY_ENHANCEMENTS: true
} as const

// === BROWSER SUPPORT ===
export const BROWSER_SUPPORT = {
  BACKDROP_FILTER: ['Chrome 88+', 'Firefox 103+', 'Safari 14+', 'Edge 88+'],
  CSS_GRID: ['Chrome 57+', 'Firefox 52+', 'Safari 10.1+', 'Edge 16+'],
  CUSTOM_PROPERTIES: ['Chrome 49+', 'Firefox 31+', 'Safari 9.1+', 'Edge 15+'],
  INTERSECTION_OBSERVER: ['Chrome 51+', 'Firefox 55+', 'Safari 12.1+', 'Edge 17+'],
  RESIZE_OBSERVER: ['Chrome 64+', 'Firefox 69+', 'Safari 13.1+', 'Edge 79+']
} as const

// === ACCESSIBILITY ===
export const ACCESSIBILITY_FEATURES = {
  REDUCED_MOTION_SUPPORT: true,
  HIGH_CONTRAST_MODE: true,
  SCREEN_READER_COMPATIBLE: true,
  KEYBOARD_NAVIGATION: true,
  FOCUS_MANAGEMENT: true,
  SEMANTIC_HTML: true
} as const 