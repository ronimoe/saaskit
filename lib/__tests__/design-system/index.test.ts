import * as DesignSystem from '@/lib/design-system/index';

describe('Design System Entry Point', () => {
  it('should export all key components and utilities', () => {
    // Components
    expect(DesignSystem.GlassCard).toBeDefined();
    expect(DesignSystem.UnfoldableFeatureCard).toBeDefined();

    // Hooks
    expect(DesignSystem.useMagneticEffect).toBeDefined();
    expect(DesignSystem.useThemeConfig).toBeDefined();

    // Utilities
    expect(DesignSystem.createBrandPalette).toBeDefined();
    expect(DesignSystem.calculateUserLevel).toBeDefined();
    
    // Tokens & Composition
    expect(DesignSystem.DESIGN_TOKENS).toBeDefined();
    expect(DesignSystem.composeComplete).toBeDefined();
  });

  it('DESIGN_SYSTEM_VERSION should match snapshot', () => {
    expect(DesignSystem.DESIGN_SYSTEM_VERSION).toMatchSnapshot();
  });

  it('FEATURES constant should match snapshot', () => {
    expect(DesignSystem.FEATURES).toMatchSnapshot();
  });

  it('BROWSER_SUPPORT constant should match snapshot', () => {
    expect(DesignSystem.BROWSER_SUPPORT).toMatchSnapshot();
  });

  it('ACCESSIBILITY_FEATURES constant should match snapshot', () => {
    expect(DesignSystem.ACCESSIBILITY_FEATURES).toMatchSnapshot();
  });
}); 