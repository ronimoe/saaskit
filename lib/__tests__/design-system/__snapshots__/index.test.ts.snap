// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design System Entry Point ACCESSIBILITY_FEATURES constant should match snapshot 1`] = `
{
  "FOCUS_MANAGEMENT": true,
  "HIGH_CONTRAST_MODE": true,
  "KEYBOARD_NAVIGATION": true,
  "REDUCED_MOTION_SUPPORT": true,
  "SCREEN_READER_COMPATIBLE": true,
  "SEMANTIC_HTML": true,
}
`;

exports[`Design System Entry Point BROWSER_SUPPORT constant should match snapshot 1`] = `
{
  "BACKDROP_FILTER": [
    "Chrome 88+",
    "Firefox 103+",
    "Safari 14+",
    "Edge 88+",
  ],
  "CSS_GRID": [
    "Chrome 57+",
    "Firefox 52+",
    "Safari 10.1+",
    "Edge 16+",
  ],
  "CUSTOM_PROPERTIES": [
    "Chrome 49+",
    "Firefox 31+",
    "Safari 9.1+",
    "Edge 15+",
  ],
  "INTERSECTION_OBSERVER": [
    "Chrome 51+",
    "Firefox 55+",
    "Safari 12.1+",
    "Edge 17+",
  ],
  "RESIZE_OBSERVER": [
    "Chrome 64+",
    "Firefox 69+",
    "Safari 13.1+",
    "Edge 79+",
  ],
}
`;

exports[`Design System Entry Point DESIGN_SYSTEM_VERSION should match snapshot 1`] = `"1.0.0"`;

exports[`Design System Entry Point FEATURES constant should match snapshot 1`] = `
{
  "ACCESSIBILITY_ENHANCEMENTS": true,
  "DYNAMIC_THEMING": true,
  "GAMIFICATION": true,
  "GLASSMORPHISM": true,
  "INTERACTIVE_CONNECTIONS": true,
  "MAGNETIC_EFFECTS": true,
  "NON_TRADITIONAL_LAYOUTS": true,
  "PARTICLE_ANIMATIONS": true,
}
`;
